### ISSUE TITLE
Evolution for the Dart 4.0 proposal

### ISSUE THREAD

**Issue by Ing-Brayan-Martinez** [dart-lang/language#3727]
# Introduction

Greetings, it's been a while since I made my [controversial proposal,](https://github.com/dart-lang/language/issues/3617) I also made a [proposal to the tensorflow team](https://github.com/tensorflow/flutter-tflite/issues/209) and I've had time to read the responses and I thought about leaving things as they were, but something happened that changed everything.

I recently read [the news that Google unified various teams](https://www.theverge.com/2024/4/18/24133881/google-android-pixel-teams-reorg-rick-osterloh) to be more competitive, this is the right decision, it seems to me that it is a necessary and important change, based on this I am going to update my proposal to make it more precise, clear and attainable because I will only talk about products under Google control.

# Proposal for Google's hardware ecosystem

It is no secret that Google was a pioneer in the creation of hardware specialized in artificial intelligence with the famous [TPU](https://cloud.google.com/blog/products/ai-machine-learning/what-makes-tpus-fine-tuned-for-deep-learning), Google is also a manufacturer of mobile phones thanks to the [Pixel family](https://store.google.com/es/category/phones?hl=es) and with all that experience they are also a manufacturer of processors but for example the [Tensor G3 ](https://www.androidpolice.com/google-tensor-g3/), so Google can be called at this time as a hardware manufacturing company. This is very important based on this context. I propose that Google concentrate on creating a family of processors that is the basis for consolidating the ecosystem of devices manufactured by Google. it should be noted that the specifications are projections of what could be, they are not exact numbers because it is a proposal to take precautions, you will make the best decision about what said family of processors will be like, knowing that, let's see the following:

- **Chip for servers:** As I mentioned before, Google is the manufacturer of [TPUs](https://cloud.google.com/tpu/docs/v5p?hl=es-419), which is perfect. My idea is to look for the evolution of this concept and think of a chip with ARM architecture that contains a large number of computing units for AI, that is, a powerful NPU. , also with a small integrated GPU possibly with 8 computing units and an integrated CPU also with 8 computing units and with 128 GB of RAM, the idea is to create a high-power silicon for industrial use, they could be inspired by Nvidia Blawerd or Intel Gaudy Then everything will be linked, let's continue.
- **Chip for PC:** Google is also a PC manufacturer thanks to the Chrome Book, so think about creating a chip with ARM architecture which has 20 NPU computing units, which are needed to move light AI models, 40 GPU computing units enough to play video games and 8 CPU computing units with 16 or 32 GB of ram would be the specifications of an average PC for working and studying, sometimes I think that the Chrome Book should evolve into the Pixel Book and be part of a product family only.
- **Chip for Smart TV:** This chip would be similar to the chip for PC so there is not much to discuss
- **Chip for Smartphones:** There is not much to say here because Google has already done an excellent job with the creation of the Tensor G3 which is used by Pixel phones.
- **Smart watch chip:** Google is also a smart watch manufacturer thanks to the Pixel Smart Watch, so it is thinking about creating a chip with ARM architecture that has 20 NPU computing units, necessary to move light AI models, 8 GPU computing, 8 CPU computing units. with 8 GB of ram.
- **Chip for Edge devices:** Google is also a manufacturer of [smart home devices,](https://store.google.com/es/category/connected_home?hl=es) so they will need a chip for these devices, a chip similar to the one used in Pixel Smart Watches could be used.

The first generation of devices made with this family of processors will be key because they will be the benchmark for long-term support

# Proposal for Google operating system

As everyone knows, Google's operating system par excellence is [Android](https://www.android.com/), so if we talk about hardware I am obliged to talk about operating systems. My proposals are to evolve Android into a family of operating systems as follows:

- **Android Server:** It would be the Android version focused on working in the cloud. The idea is to create an operating system that makes the most of the TPUs, which are specialized hardware for artificial intelligence. Currently, I can imagine that they make use of a modified Linux distribution to run on the TPUs. Yes. We already have this, it will only be a matter of Android working with this modified Linux distribution since Android is also based on the Linux kernel android must be a general purpose system
- **Android Desktop:** It would be the Android version focused on working on PC, it would be the version for professional use of Android and it would be the substitute for Chrome OS. I think that the team that maintains ChromeOS should create Android Desktop because it already has experience and is a valuable resource. Android Desktop It should be compatible with laptops with ARM chip manufactured by Google
- **Android TV:** It is the Android version focused on smart TVs. This version already exists and is a successful product, so I have nothing to say except that it is compatible with the TV chips manufactured by Google.
- **Android:** It is the original Android version focused on smart phones, this version already exists and is a successful product so I have nothing to say except that it continues to evolve thanks to Pixel phones
- **Android Watch**: It is the original Android version focused on smart watches. This version already exists and is a successful product, so I have nothing to say except that it continues to evolve thanks to the Pixel Smart Watch phones.
- **Android for IoT:** It is the original Android version focused on smart home devices, this version already exists and is a successful product so I have nothing to say except that it continues to evolve

Thanks to this entire family of operating systems, Google will be able to create a solid ecosystem that can take full advantage of the hardware manufactured by Google for all NPUs. Afterwards, with agreements with other companies such as Intel, thanks to the [uxl foundation](https://uxlfoundation.org/), it will be possible to think of a compatible family for X86 architecture but the important thing is the family of Android systems for ARM

# Proposals for Dart as a language for AI

Already with the ecosystem of devices proposed above, now if I am going to talk about the Dart language, as I mentioned before, I think that Dart is a competent alternative to Python because it is an excellent language with a mature syntax and good interoperability with C++ Rust and Objective C code.

In addition, Google has an excellent artificial intelligence framework such as [Tensorflow](https://www.tensorflow.org/?hl=es-419), which can be thought of as a combination between Dart and C++ that is compatible with the entire family of chips proposed previously and takes advantage of all the NPUs available on different platforms. I said this in the Past proposal AI and multiplatform should be Google's priority

The idea is that lightweight AI models such as:

- Gemini Nano
- Gemma
- Llama 3 8B
- Mistral 7B Instruck
- Phi 3
- Claude 3 Haiku
- Among others

All these lightweight AI models will be very important for creating applications and providing good support as a priority for Google.

# Proposals for Dart and Flutter

As we all know, [Flutter](https://flutter.dev/) is the main framework of the Dart ecosystem and it is obvious that it will also be affected by all these proposals.

From my point of view, Flutter is a fairly complete framework. The changes that I see are the support for applications for the family of Android operating systems that has been proposed previously so that with the same code base you can quickly create software solutions in multiple applications. devices

With the help of the Tensorflow version made with Dart, all applications made with Flutter will be able to run AI models locally, taking full advantage of the hardware created by Google, making the most of all the properties of the Dart language and its excellent interportability with C++

# Proposal for Dart as a video game development language

As we all know, Flutter is a framework that uses a graphics engine called Impeller to render applications. This is possible thanks to the excellent interoperability with C++. Flutter in combination with [Flame](https://flame-engine.org/) is a base platform for creating cross-platform video games.

My proposal is to go a little further, and collaborate with the open source Godot project, which is a graphics engine for developing video games and is being very popular lately.

[Godot](https://godotengine.org/) being an open source project, it is possible to intervene and improve it, so my proposal is based on supporting the Dart language so that it works with Godot's echo C++ core and that the Dart language is an alternative Gdscript language that is a bit Achilles heel of this technology, the support of the Dart language is possible thanks to its excellent interoperability with C++

The development of the Impeller graphics engine is an important investment and its evolution is essential so that the video games developed with Dart are of higher quality.

Another important point thanks to the family of processors mentioned above equipped with NPUs, the use of AI models that run locally will be essential to create image rescaling very similar to what Nvidia DLSS or Intel XeSS or Apple's MetalFX does. of tensorflow in this aspect is very important to develop the Dart language as a language for video games

# Conclusions

The idea of all these proposals is to give a vision as a reference point of how to evolve the Dart language, all this is with a long-term vision, the teams involved will go step by step building results. I hope to be able to explain myself, I just want to collaborate with my favorite language

Try to be as precise and minimalist as possible with achievable goals and that they are under the control of Google, the last word is yours that it is possible and that it is not. Greetings

**Comment by mraleph** [comment]
I really encourage you to understand that this is the wrong place for these. Maybe you could create a blog and post these there.
